adds chart for kafka

faraazkhan opened the pull request:
Adds a chart for Kafka based on the awesome work done here: https://github.com/Yolean/kubernetes-kafka and here: https://github.com/kubernetes/charts/tree/master/incubator/zookeeper

faraazkhan committed:
adds chart for kafka

prydonius commented:
@k8s-bot ok to test

prydonius commented:
@faraazkhan can you sign the Linux Foundation CLA?

faraazkhan commented:
@prydonius I am trying. Added my github account and everything,  it says, I am supposed to get an email for with the CLA but nothing so far.

faraazkhan commented:
Also looks like we are running an old version of helm in Jenkins right now. `I1021 19:04:55.614] [ERROR] templates/kafka-petset.yaml: namespace option is currently NOT supported
` doesnt seem right 😄 

Update: Verified Jenkins is running a newer version `Client: &version.Version{SemVer:"v2.0.0-beta.1", GitCommit:"1b792f990675702488899b94352168572c54c972", GitTreeState:"clean"}` . I have tested this against `Client: &version.Version{SemVer:"v2.0.0-alpha.5", GitCommit:"a324146945c01a1e2dd7eaf23caf0e55fabfd3d2", GitTreeState:"clean"}`

Not sure why the error then 😞

prydonius commented:
@faraazkhan Thanks for signing the CLAs! Namespace shouldn't be set in the manifests, since Helm manages which namespace a Chart's resources gets installed into using the `helm install --namespace` flag.

prydonius commented on incubator/kafka/Chart.yaml:
```diff
@@ -0,0 +1,16 @@
+home: https://kafka.apache.org/
+sources:
+  - https://github.com/kubernetes/charts/tree/master/incubator/zookeeper
+  - https://github.com/Yolean/kubernetes-kafka
```
can you add `https://github.com/solsson/dockerfiles/tree/master/kafka`

prydonius commented on incubator/kafka/charts/zookeeper/Chart.yaml:
```diff
@@ -0,0 +1,10 @@
+name: zookeeper
```
Helm now supports specifying dependencies using a `requirements.yaml` file. Can you update this to remove this vendored version of zookeeper, and instead just add a requirements.yaml that can be used to bring in the dependency? 

See https://github.com/kubernetes/helm/blob/master/docs/charts.md#managing-dependencies-with-requirementsyaml and https://github.com/kubernetes/charts/tree/master/incubator/patroni for an example.

prydonius commented on incubator/kafka/values.yaml:
```diff
@@ -0,0 +1,30 @@
+global:
+  Namespace: "kafka"
+  ZooKeeperServiceName: "zookeeper"
+
```
Can you remove these trailing lines?

faraazkhan committed:
upgrading chart to using requirements.yaml for dependencies

faraazkhan committed:
updates README.md

faraazkhan commented:
@prydonius Jenkins is currently failing with `W1023 20:38:31.781] Error: deletion error count 1: no objects visited`. I have seen this error, along with at least one more in the latest version of helm. This doesn't necessarily seem like an issue with the chart, but more with helm?

faraazkhan commented:
@k8s-bot e2e test this

chrislovecnm commented on incubator/kafka/README.md:
```diff
@@ -0,0 +1,111 @@
+
+## Pre Requisites:
+
+* Kubernetes 1.3 with alpha APIs enable
```
You are also using storage classes as well.  You probably should mention both of those.

chrislovecnm commented on incubator/kafka/README.md:
```diff
@@ -0,0 +1,111 @@
+
+### Connecting to Kafka
+
+You can connect to Kafka by running a simple pod in the K8s cluster like this with a configuration like this:
```
You may want to add a notes.txt file with these instructions.

chrislovecnm commented on incubator/kafka/templates/kafka-petset.yaml:
```diff
@@ -0,0 +1,94 @@
+        - containerPort: 9092
+          name: kafka
+        resources:
+          requests:
```
We need a limits section in here as well.

chrislovecnm commented on incubator/kafka/templates/kafka-petset.yaml:
```diff
@@ -0,0 +1,94 @@
+          mountPath: /tmp/zookeeper
+      volumes:
+      - name: workdir
+        emptyDir: {}
```
Dumb question but does this need persitence?

chrislovecnm commented on incubator/kafka/templates/kafka-petset.yaml:
```diff
@@ -0,0 +1,94 @@
+  - metadata:
+      name: datadir
+      annotations:
+        volume.alpha.kubernetes.io/storage-class: anything
```
Should we have this setup to do both a PVC or an empty dir?  Looking for an example ...

faraazkhan replied:
I think the empty dir can be removed. Want to leave the PVC in. I will push that out shortly.

chrislovecnm commented on incubator/kafka/templates/kafka-petset.yaml:
```diff
@@ -0,0 +1,94 @@
+        pod.alpha.kubernetes.io/initialized: "true"
+        pod.alpha.kubernetes.io/init-containers: '[
+        ]'
+    spec:
```
We need health check probes http://kubernetes.io/docs/user-guide/liveness/

chrislovecnm replied:
Still needs a ready probe.  We can move into incubator without.

faraazkhan replied:
I could just add the same test for readiness as well. Not sure how valuable that would be though, in Kafka's situation.

chrislovecnm reviewed (changes requested):
Multiple tweaks ... looks good

chrislovecnm commented:
@viglesiasce any ideas on the build failure "W1023 20:38:31.781] Error: deletion error count 1: no objects visited", do we need to bump the helm version we are using?

krisnova commented on incubator/kafka/.helmignore:
```diff
@@ -0,0 +1,21 @@
+*.bak
+*.tmp
+*~
+# Various IDEs
```
Is it best practice to add proprietary file types to a global `.ignore` file? Shouldn't the user be managing these?

chrislovecnm replied:
@kris-nova this is autogenned by helm.

faraazkhan committed:
adds notes.txt

Adds liveness probe
Adds resource limits
Improves README.md

faraazkhan committed:
Merge remote-tracking branch 'origin/master' into kafka

faraazkhan commented:
Ok. I have addressed the comments here . But still getting the `Error: deletion error count 1: no objects visited` on Jenkins. I also see this happen locally when I do a `helm delete --purge my-release`. It seems to be raised [here](https://github.com/kubernetes/helm/blob/846011b02abc600e5b21a88f3b407f46e79cdc81/pkg/kube/client.go#L316), will try to dig into it further later, in the meanwhile any help is appreciated.  @prydonius @viglesiasce @chrislovecnm 

Update: Please note, despite the error all the resources seem to be deleted fine and I am able to reuse the helm release name even in the same namespace with no errors (so purge worked just fine)

chrislovecnm commented:
@k8s-bot e2e test this

chrislovecnm reviewed (commented):
Still, have storage class issue, and readyprobe not defined, and we can merge this into incubator with these open items.

@viglesiasce any ideas about the build failures?

Thanks for the chart @faraazkhan!!!

faraazkhan committed:
adds readiness probe and removes empty dir

viglesiasce commented:
Sorry for the delay @faraazkhan! Will look into the failures today.

technosophos commented:
With that deletion error, can you still see the resources in Kubernetes that were supposed to be deleted? That error message usually indicates that the deletion logic went to remove a resource from Kubernetes, and the resource was missing.

viglesiasce commented:
One thing I noticed is that after deploy the kf and kf-broker services end up with no endpoints.

@faraazkhan is that expected?

@technosophos could that be an issue?

``` console
NAME                        ENDPOINTS                                                        AGE
ep/demo-app-promchart       10.140.5.17:8080,10.140.5.18:8080,10.140.5.25:8080 + 2 more...   1d
ep/demo-grafana             10.140.5.21:3000                                                 1d
ep/kubernetes               104.199.112.125:443                                              16d
ep/plinking-racoo-alerts    10.140.5.6:9093                                                  5d
ep/plinking-racoo-server    10.140.5.7:9090                                                  5d
ep/steely-panda-nginx       10.140.5.23:80                                                   21h
ep/voting-goose-kf          <none>                                                           5m
ep/voting-goose-kf-broker   <none>                                                           5m
ep/voting-goose-zk          10.140.1.11:2888,10.140.1.13:2888,10.140.1.14:2888 + 3 more...   5m
NAME                         CLUSTER-IP       EXTERNAL-IP   PORT(S)             AGE
svc/demo-app-promchart       10.143.244.167   <none>        80/TCP              1d
svc/demo-grafana             10.143.246.48    <none>        80/TCP              1d
svc/kubernetes               10.143.240.1     <none>        443/TCP             16d
svc/plinking-racoo-alerts    10.143.245.26    <none>        9093/TCP            5d
svc/plinking-racoo-server    10.143.244.45    <none>        80/TCP              5d
svc/steely-panda-nginx       10.143.242.164   <none>        80/TCP              21h
svc/voting-goose-kf          10.143.243.96    <none>        9092/TCP            5m
svc/voting-goose-kf-broker   None             <none>        9092/TCP            5m
svc/voting-goose-zk          None             <none>        2888/TCP,3888/TCP   5m
```

chrislovecnm commented:
Houston that is a problem  with the services ;) @faraazkhan did you monkey your selectors or @viglesiasce do yah think this is helm?

faraazkhan committed:
ensure endpoints for all services and better testing instructions

faraazkhan commented:
@chrislovecnm @viglesiasce @technosophos sorry about the delay in responding, so I just pushed a few changes, that ensure that all the services created have endpoints. I see what @technosophos is saying but not sure what that resource could be, everything created in the helm chart seems to be destroyed correctly:

![nov-09-2016 20-38-24](https://cloud.githubusercontent.com/assets/613236/20162736/93459c62-a6bc-11e6-8d13-34f8cc0ff3d3.gif)

viglesiasce commented:
Thanks @faraazkhan! Things are looking better wrt services. Can you add a NOTES file that explains next steps to the users.

faraazkhan commented:
@viglesiasce Can you please be specific, how can I improve this? Maybe I am not following a convention `notes` vs `NOTES` https://github.com/kubernetes/charts/pull/144/files#diff-2654b4aa4eacf071f2e90debaad8b050R1

chrislovecnm commented:
NOTES.txt file. Look at the charts that are not in incubator. They should have examples

prydonius commented on incubator/kafka/notes.txt:
```diff
@@ -0,0 +1,36 @@
+You can connect to Kafka by running a simple pod in the K8s cluster like this with a configuration like this:
```
Helm will specifically look for a while called `NOTES.txt`, render it as a template, and print it out after installing/upgrading or running `helm status` on the release. This will allow you to print out working commands with the name of the release! See  https://github.com/kubernetes/charts/blob/master/incubator/patroni/templates/NOTES.txt for an example :)

chrislovecnm replied:
:thumbsup

chrislovecnm commented on incubator/kafka/README.md:
```diff
@@ -0,0 +1,113 @@
+
+## Known Limitations
+
+* Namespace creation is not automated
```
This is an easy fix. Will need to happen for stable.  This is may be a reason why unit tests are failing.

viglesiasce replied:
@faraazkhan any word on what these services are for?

chrislovecnm reviewed (changes requested):
This is also and issue

``` console
oot@ip-172-20-82-63:/var/log# kubectl -n kafka describe svc
Name:           exacerbated-ar-kafka
Namespace:      kafka
Labels:         chart=kafka-0.1.0
            component=exacerbated-ar-kafka
            heritage=Tiller
            release=exacerbated-ar
Selector:       app=kafka,component=exacerbated-ar-kafka
Type:           ClusterIP
IP:         100.69.243.160
Port:           <unset> 9092/TCP
Endpoints:      <none>
Session Affinity:   None
No events.

Name:           exacerbated-ar-kafka-bro
Namespace:      kafka
Labels:         chart=kafka-0.1.0
            component=exacerbated-ar-kafka
            heritage=Tiller
            release=exacerbated-ar
Selector:       app=kafka,component=exacerbated-ar-kafka
Type:           ClusterIP
IP:         None
Port:           <unset> 9092/TCP
Endpoints:      <none>
Session Affinity:   None
No events.

Name:           exacerbated-ar-zk
Namespace:      kafka
Labels:         app=zk
            chart=zookeeper-0.1.1
            component=exacerbated-ar-zk
            heritage=Tiller
            release=exacerbated-ar
Selector:       app=zk,component=exacerbated-ar-zk
Type:           ClusterIP
IP:         None
Port:           client  2181/TCP
Endpoints:      100.96.3.3:2181,100.96.4.5:2181,100.96.5.2:2181
Port:           peer    2888/TCP
Endpoints:      100.96.3.3:2888,100.96.4.5:2888,100.96.5.2:2888
Port:           leader-election 3888/TCP
Endpoints:      100.96.3.3:3888,100.96.4.5:3888,100.96.5.2:3888
Session Affinity:   None
No events.

Name:           zookeeper
Namespace:      kafka
Labels:         <none>
Selector:       app=zk
Type:           ClusterIP
IP:         100.65.108.22
Port:           client  2181/TCP
Endpoints:      100.96.3.3:2181,100.96.4.5:2181,100.96.5.2:2181
Session Affinity:   None
```

You have two services that do not have endpoints.  I am uncertain what the purpose of the services are.

viglesiasce commented:
@k8s-bot ok to test

viglesiasce commented on incubator/kafka/notes.txt:
```diff
@@ -0,0 +1,36 @@
+  kind: Pod
+  metadata:
+    name: testclient
+    namespace: kafka
```
I wouldnt set the namespace explicitly here. Users may not have configured this namespace for installation. 

You can make this a kubectl run command and avoid having the pod yaml here. Look at the postgresql chart as an example.

`kubectl run client -it --rm --image solsson/kafka:0.10.0.1 --command -- bash`

viglesiasce commented on incubator/kafka/notes.txt:
```diff
@@ -0,0 +1,36 @@
+  To stop the listener session above press: Ctrl+C
+
+  To start an interactive message producer session:
+  kubectl -n kafka exec -ti testclient -- ./bin/kafka-console-producer.sh --broker-list my-release-broker-kf.kafka.svc.cluster.local:9092 --topic test1
```
You can simplify the dns name there and make it `--broker-list {{ .Release.Name }}-kf:9092`. Again this make the command work across namespaces.

viglesiasce reviewed (changes requested):
1) Move notes.txt to templates/NOTES.txt
2) Update NOTES.txt to be templatized
3) Remove dependencies in notes on a particular namespace

F21 commented:
Any update here?

prydonius commented:
@faraazkhan any updates here?
